{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial 1 – Import a GEM, Set Parameters and Run FBA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a short introduction that shows how to load a genome-scale metabolic model (GEM), set reaction constraints, objective function and perform an optimization through flux balance analysis (FBA). The resulting fluxes are visualized and exported to a PDF file.\n",
    "\n",
    "A GEM for the filamentous fungus Penicillium chrysogenum is used in this tutorial. The model can be found in a Microsoft Excel file under the name iAL1006 v1.00.xlsx and in SBML file iAL1006 v1.00.xml.\n",
    "Open tutorial1.m file in MATLAB to begin this exercise. To run a section of code in MATLAB, highlight it, press right mouse button on it and choose an option “Evaluate selection”.\n",
    "\n",
    "NOTE: the user must be able to successfully import the GEMs in Excel format with the RAVEN function importExcelModel. Although, this functionality is not necessary for this exercise, the users without such ability would not be able to do Tutorials 2-4, which involve working with GEMs in RAVEN compatible Excel format.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "setRavenSolver('gurobi')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "- [importExcelModel](https://sysbiochalmers.github.io/RAVEN/doc/io/importExcelModel.html)\n",
    "\n",
    "```MATLAB\n",
    "function model=importExcelModel(fileName,removeExcMets,printWarnings,ignoreErrors)\n",
    "```\n",
    "Imports a constraint-based model from a Excel file\n",
    "\n",
    "- [importModel](https://sysbiochalmers.github.io/RAVEN/doc/io/importModel.html)\n",
    "\n",
    "```MATLAB\n",
    "function model=importModel(fileName,removeExcMets,isSBML2COBRA,supressWarnings)\n",
    "```\n",
    "Import a constraint-based model from a SBML file\n",
    "\n",
    "---\n",
    "\n",
    "> removeExcMets: true if exchange metabolites should be removed. This is needed to be able to run simulations, but it could also be done using simplifyModel at a later stage (opt, default true)\"\n",
    "\n",
    "The tutorial file explains:<br>\n",
    "*\"The \"false\" flag imports a model with exchange reactions in their \"closed\" form. This makes the model unsuited for modelling, but it is useful for some quality control steps.\"*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The model contains 47 errors.\n",
      "\n",
      "Error encountered during read.\n",
      "\n",
      "WARNING: No objective function found. This might be intended, but results in FBCv2 non-compliant SBML file when exported\n",
      "\n",
      "WARNING: The composition for the following metabolites could not be parsed:\n",
      "\tLPE\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<html><body><pre>model = struct with fields:\n",
       "   id: 'iAL1006'\n",
       "              name: 'Penicillium chrysogenum genome-scale model'\n",
       "        annotation: [1x1 struct]\n",
       "              rxns: {1632x1 cell}\n",
       "              mets: {1395x1 cell}\n",
       "                 S: [1395x1632 double]\n",
       "                lb: [1632x1 double]\n",
       "                ub: [1632x1 double]\n",
       "               rev: [1632x1 double]\n",
       "                 c: [1632x1 double]\n",
       "                 b: [1395x1 double]\n",
       "             comps: {5x1 cell}\n",
       "         compNames: {5x1 cell}\n",
       "       compMiriams: {5x1 cell}\n",
       "          rxnNames: {1632x1 cell}\n",
       "          rxnComps: [1632x1 double]\n",
       "           grRules: {1632x1 cell}\n",
       "        rxnGeneMat: [1632x1006 double]\n",
       "        subSystems: {1632x1 cell}\n",
       "           eccodes: {1632x1 cell}\n",
       "        rxnMiriams: {1632x1 cell}\n",
       "          rxnNotes: {1632x1 cell}\n",
       "     rxnReferences: {1632x1 cell}\n",
       "             genes: {1006x1 cell}\n",
       "       geneMiriams: {1006x1 cell}\n",
       "    geneShortNames: {1006x1 cell}\n",
       "          metNames: {1395x1 cell}\n",
       "          metComps: [1395x1 double]\n",
       "            inchis: {1395x1 cell}\n",
       "       metFormulas: {1395x1 cell}\n",
       "        metMiriams: {1395x1 cell}\n",
       "     unconstrained: [1395x1 double]</pre></body></html>"
      ],
      "text/plain": [
       "model = struct with fields:\n",
       "   id: 'iAL1006'\n",
       "              name: 'Penicillium chrysogenum genome-scale model'\n",
       "        annotation: [1x1 struct]\n",
       "              rxns: {1632x1 cell}\n",
       "              mets: {1395x1 cell}\n",
       "                 S: [1395x1632 double]\n",
       "                lb: [1632x1 double]\n",
       "                ub: [1632x1 double]\n",
       "               rev: [1632x1 double]\n",
       "                 c: [1632x1 double]\n",
       "                 b: [1395x1 double]\n",
       "             comps: {5x1 cell}\n",
       "         compNames: {5x1 cell}\n",
       "       compMiriams: {5x1 cell}\n",
       "          rxnNames: {1632x1 cell}\n",
       "          rxnComps: [1632x1 double]\n",
       "           grRules: {1632x1 cell}\n",
       "        rxnGeneMat: [1632x1006 double]\n",
       "        subSystems: {1632x1 cell}\n",
       "           eccodes: {1632x1 cell}\n",
       "        rxnMiriams: {1632x1 cell}\n",
       "          rxnNotes: {1632x1 cell}\n",
       "     rxnReferences: {1632x1 cell}\n",
       "             genes: {1006x1 cell}\n",
       "       geneMiriams: {1006x1 cell}\n",
       "    geneShortNames: {1006x1 cell}\n",
       "          metNames: {1395x1 cell}\n",
       "          metComps: [1395x1 double]\n",
       "            inchis: {1395x1 cell}\n",
       "       metFormulas: {1395x1 cell}\n",
       "        metMiriams: {1395x1 cell}\n",
       "     unconstrained: [1395x1 double]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "% How to load a model\n",
    "model=importModel('tutorial_data/iAL1006 v1.00.xml',false)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "%The following function prints some properties of the model. The two \"true\"\n",
    "%flags say that it should also list potential problems such as dead-end\n",
    "%reactions or unconnected metabolites."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [printModelStats](https://sysbiochalmers.github.io/RAVEN/doc/core/printModelStats.html)\n",
    "\n",
    "```MATLAB\n",
    "function printModelStats(model, printModelIssues, printDetails)\n",
    "```\n",
    "\n",
    "Prints some statistics about a model to the screen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Network statistics for iAL1006: Penicillium chrysogenum genome-scale model\n",
      "EC-numbers\t\t\t626\n",
      "Genes*\t\t\t\t1006\n",
      "\tPeroxisome\t38\n",
      "\tMitochondria\t219\n",
      "\tCytosol\t802\n",
      "\tExtracellular\t95\n",
      "\tBoundary\t0\n",
      "\n",
      "Reactions*\t\t\t1632\n",
      "\tPeroxisome\t137\n",
      "\tMitochondria\t324\n",
      "\tCytosol\t1144\n",
      "\tExtracellular\t336\n",
      "\tBoundary\t161\n",
      "Unique reactions**\t1449\n",
      "\n",
      "Metabolites\t\t\t1395\n",
      "\tPeroxisome\t105\n",
      "\tMitochondria\t242\n",
      "\tCytosol\t728\n",
      "\tExtracellular\t160\n",
      "\tBoundary\t160\n",
      "Unique metabolites\t849\n",
      "\n",
      "* Genes and reactions are counted for each compartment if any of the corresponding metabolites are in that compartment. The sum may therefore not add up to the total number.\n",
      "** Unique reactions are defined as being biochemically unique (no compartmentalization)\n",
      "\n",
      "Short model quality summary for iAL1006: Penicillium chrysogenum genome-scale model\n",
      "Dead-end reactions\t66\n",
      "\tr0035\n",
      "\tr0036\n",
      "\tr0040\n",
      "\tr0061\n",
      "\tr0071\n",
      "\tr0108\n",
      "\tr0161\n",
      "\tr0162\n",
      "\tr0169\n",
      "\tr0170\n",
      "\tr0174\n",
      "\tr0213\n",
      "\tr0224\n",
      "\tr0235\n",
      "\tr0238\n",
      "\tr0240\n",
      "\tr0356\n",
      "\tr0405\n",
      "\tr0407\n",
      "\tr0424\n",
      "\tr0427\n",
      "\tr0429\n",
      "\tr0447\n",
      "\tr0454\n",
      "\tr0514\n",
      "\tr0515\n",
      "\tr0516\n",
      "\tr0533\n",
      "\tr0534\n",
      "\tr0535\n",
      "\tr0538\n",
      "\tr0539\n",
      "\tr0573\n",
      "\tr0576\n",
      "\tr0577\n",
      "\tr0580\n",
      "\tr0581\n",
      "\tr0582\n",
      "\tr0586\n",
      "\tr0597\n",
      "\tr0602\n",
      "\tr0603\n",
      "\tr0604\n",
      "\tr0612\n",
      "\tr0630\n",
      "\tr0631\n",
      "\tr0641\n",
      "\tr0691\n",
      "\tr0697\n",
      "\tr0707\n",
      "\tr0712\n",
      "\tr0723\n",
      "\tr0726\n",
      "\tr0730\n",
      "\tr0731\n",
      "\tr0732\n",
      "\tr0752\n",
      "\tr0754\n",
      "\tr1166\n",
      "\tr1221\n",
      "\tr1275\n",
      "\tr0234\n",
      "\tr0428\n",
      "\tr0437\n",
      "\tr0517\n",
      "\tr0537\n",
      "\n",
      "Dead-end metabolites\t232\n",
      "\t(LLACb) (S)-lactate\n",
      "\t(MALb) (S)-malate\n",
      "\t(13GLUCANb) 1,3-beta-D-glucan\n",
      "\t(2HPENGb) 2-hydroxybenzylpenicillin\n",
      "\t(AKGb) 2-oxoglutarate\n",
      "\t(C4ODb) 2-oxy-but-3-enoate\n",
      "\t(PABAb) 4-aminobenzoate\n",
      "\t(4HPOAb) 4-hydroxyphenoxyacetate\n",
      "\t(4HPENVb) 4-hydroxyphenoxymethylpenicillin\n",
      "\t(6APAb) 6-aminopenicillanate\n",
      "\t(OPCb) 6-oxopiperidine-2-carboxylate\n",
      "\t(8HPAb) 8-hydroxypenillic acid\n",
      "\t(ACb) acetate\n",
      "\t(ADb) adenine\n",
      "\t(TREb) alpha,alpha-trehalose\n",
      "\t(GLCb) alpha-D-glucose\n",
      "\t(ARABINb) alpha-L-arabinan\n",
      "\t(AMIACEb) aminoacetaldehyde\n",
      "\t(ANb) anthranilate\n",
      "\t(PENARTb) artificial penicillin\n",
      "\t(ARTPROTb) artificial protein\n",
      "\t(PENGb) benzylpenicillin\n",
      "\t(PENGAb) benzylpenicilloic acid\n",
      "\t(bALAb) beta-alanine\n",
      "\t(bDGLCb) beta-D-glucose\n",
      "\t(BIOMASSb) biomass\n",
      "\t(C40b) butyrate\n",
      "\t(CB15LCTb) cellobiono-1,5-lactone\n",
      "\t(CELLOBb) cellobiose\n",
      "\t(CELLUb) cellulose\n",
      "\t(CHITb) chitin\n",
      "\t(CHIBb) chitobiose\n",
      "\t(CHITOb) chitosan\n",
      "\t(CHOb) choline\n",
      "\t(CITb) citrate\n",
      "\t(CO2b) CO2\n",
      "\t(CYNEb) cyanate\n",
      "\t(CYTSb) cytosine\n",
      "\t(AOLb) D-arabinitol\n",
      "\t(ARABb) D-arabinose\n",
      "\t(C100b) decanoate\n",
      "\t(FRUb) D-fructose\n",
      "\t(GLACb) D-galactose\n",
      "\t(GLCNTb) D-gluconate\n",
      "\t(GLCN15LACb) D-glucono-1,5-lactone\n",
      "\t(GLCNb) D-glucosamine\n",
      "\t(DGLCb) D-glucose\n",
      "\t(MNTb) D-mannitol\n",
      "\t(MANb) D-mannose\n",
      "\t(RIBb) D-ribose\n",
      "\t(XYLb) D-xylose\n",
      "\t(ETHb) ethanol\n",
      "\t(ETHNITb) ethylnitronate\n",
      "\t(FMNb) FMN\n",
      "\t(FORb) formate\n",
      "\t(FUMb) fumarate\n",
      "\t(GABAb) gamma-aminobutyrate\n",
      "\t(RGTb) glutathione\n",
      "\t(GLb) glycerol\n",
      "\t(GLYBETb) glycine betaine\n",
      "\t(GLYb) glycine\n",
      "\t(GLYCOGENb) glycogen\n",
      "\t(GLYAb) glycolate\n",
      "\t(GNb) guanine\n",
      "\t(H2Ob) H2O\n",
      "\t(H2O2b) H2O2\n",
      "\t(H2Sb) H2S\n",
      "\t(C170b) heptadecanoate\n",
      "\t(C171b) heptadecenoate\n",
      "\t(C70b) heptanoate\n",
      "\t(C162b) hexadecadienoate\n",
      "\t(C161b) hexadecenoate\n",
      "\t(C60b) hexanoate\n",
      "\t(PENDFb) hexanoylpenicillin\n",
      "\t(HYXNb) hypoxanthine\n",
      "\t(C200b) icosanoate\n",
      "\t(ICITb) isocitrate\n",
      "\t(IPNb) isopenicillin N\n",
      "\t(AMAb) L-2-aminoadipate\n",
      "\t(LACTb) lactose\n",
      "\t(ALAb) L-alanine\n",
      "\t(LAOLb) L-arabinitol\n",
      "\t(LARABb) L-arabinose\n",
      "\t(ARGb) L-arginine\n",
      "\t(ASNb) L-asparagine\n",
      "\t(ASPb) L-aspartate\n",
      "\t(C120b) laurate\n",
      "\t(CITRb) L-citrulline\n",
      "\t(CYSb) L-cysteine\n",
      "\t(CYSTb) L-cystine\n",
      "\t(DOQUIb) L-dopaquinone\n",
      "\t(GLUb) L-glutamate\n",
      "\t(GLNb) L-glutamine\n",
      "\t(HISb) L-histidine\n",
      "\t(HCYSb) L-homocysteine\n",
      "\t(IDOLb) L-iditol\n",
      "\t(ILEb) L-isoleucine\n",
      "\t(LEUb) L-leucine\n",
      "\t(LYSb) L-lysine\n",
      "\t(METb) L-methionine\n",
      "\t(ORNb) L-ornithine\n",
      "\t(PHEb) L-phenylalanine\n",
      "\t(PROb) L-proline\n",
      "\t(LRLb) L-ribulose\n",
      "\t(SERb) L-serine\n",
      "\t(SORb) L-sorbose\n",
      "\t(THRb) L-threonine\n",
      "\t(TRPb) L-tryptophan\n",
      "\t(TYRb) L-tyrosine\n",
      "\t(VALb) L-valine\n",
      "\t(MLTb) maltose\n",
      "\t(MLTIOSEb) maltotriose\n",
      "\t(MANNANb) mannans\n",
      "\t(MELIb) melibiose\n",
      "\t(METHOLb) methanol\n",
      "\t(MYOIb) myo-inositol\n",
      "\t(C140b) myristate\n",
      "\t(NAGb) N-acetyl-D-glucosamine\n",
      "\t(NH3b) NH3\n",
      "\t(NICDb) nicotinamide\n",
      "\t(NICAb) nicotinate\n",
      "\t(HNO3b) nitrate\n",
      "\t(HNO2b) nitrite\n",
      "\t(C90b) nonanoate\n",
      "\t(O2b) O2\n",
      "\t(C182b) octadecadienoate\n",
      "\t(C183b) octadecatrienoate\n",
      "\t(C181b) octadecenoate\n",
      "\t(C80b) octanoate\n",
      "\t(PENKb) octanoylpenicillin\n",
      "\t(OXALb) oxalate\n",
      "\t(OAb) oxaloacetate\n",
      "\t(C160b) palmitate\n",
      "\t(C150b) pentadecanoate\n",
      "\t(POAb) phenoxyacetate\n",
      "\t(PENVb) phenoxymethylpenicillin\n",
      "\t(PENVAb) phenoxymethylpenicilloic acid\n",
      "\t(PAAb) phenylacetate\n",
      "\t(PIb) phosphate\n",
      "\t(6CARHEXb) pimelate\n",
      "\t(PROPb) propionate\n",
      "\t(PYRb) pyruvate\n",
      "\t(QUINb) quinolinate\n",
      "\t(RAFFb) raffinose\n",
      "\t(STARb) starch\n",
      "\t(C180b) stearate\n",
      "\t(SUCCb) succinate\n",
      "\t(SUCb) sucrose\n",
      "\t(SLFb) sulfate\n",
      "\t(H2SO3b) sulfite\n",
      "\t(Sb) sulfur\n",
      "\t(THMEb) thiamin\n",
      "\t(URAb) uracil\n",
      "\t(UREAb) urea\n",
      "\t(URIb) uridine\n",
      "\t(C50b) valerate\n",
      "\t(XANb) xanthine\n",
      "\t(XYLANb) xylans\n",
      "\t(XOLb) xylitol\n",
      "\t(DIDIPC) (S)-2,3-dihydrodipicolinate\n",
      "\t(ACPC) 1-aminocyclopropanecarboxylate\n",
      "\t(PINS3P) 1-phosphatidyl-1D-myo-inositol 3-phosphate\n",
      "\t(PINS4P) 1-phosphatidyl-1D-myo-inositol 4-phosphate\n",
      "\t(D45PI) 1-phosphatidyl-D-myo-inositol 4,5-bisphosphate\n",
      "\t(ACLAC) 2-acetolactate\n",
      "\t(AM6SA) 2-aminomuconic 6-semialdehyde\n",
      "\t(2D3DGALT) 2-dehydro-3-deoxy-D-galactonate\n",
      "\t(DEXG) 2-deoxy-D-gluconic acid\n",
      "\t(DEORIPI) 2-deoxy-D-ribose 5-phosphate\n",
      "\t(OXGLY) 2-hydroxy-3-oxosuccinic acid\n",
      "\t(KEMYOI) 2-inosose\n",
      "\t(PHEETHAL) 2-phenylethanol\n",
      "\t(DEHXG) 3-dehydro-2-deoxy-D-gluconate\n",
      "\t(APEBU) 4-(2-Aminophenyl)-2,4-dioxobutanoate\n",
      "\t(AHMPP) 4-amino-2-methyl-5-diphosphomethylpyrimidine\n",
      "\t(AHMP) 4-amino-2-methyl-5-phosphomethylpyrimidine\n",
      "\t(THZP) 4-methyl-5-(2-phosphonooxyethyl)thiazole\n",
      "\t(THZ) 5-(2-hydroxyethyl)-4-methylthiazole\n",
      "\t(ADPR) 5-deoxy-D-ribofuranos-5-yl-ADP\n",
      "\t(MTHPTGLU) 5-methyltetrahydropteroyltri-L-glutamate\n",
      "\t(ACTP) acetyl phosphate\n",
      "\t(AGMT) agmatine\n",
      "\t(GLUCRE) beta-D-glucosiduronic acids\n",
      "\t(CINNAM) cinnamate\n",
      "\t(ARABLAC) D-arabinono-1,4-lactone\n",
      "\t(DC) deoxycytidine\n",
      "\t(DU) deoxyuridine\n",
      "\t(GLCUNT) D-glucuronate\n",
      "\t(DPRO) D-proline\n",
      "\t(TDP) D-tagatofuranose 1,6-bisphosphate\n",
      "\t(DTDP) dTDP\n",
      "\t(DTTP) dTTP\n",
      "\t(GALOL) galactitol\n",
      "\t(PPGPP) guanosine 3',5'-bis(diphosphate)\n",
      "\t(PPPGPP) guanosine 3'-diphosphate 5'-triphosphate\n",
      "\t(IAD) indole-3-acetamide\n",
      "\t(IAC) indole-3-acetate\n",
      "\t(ACNL) indole-3-acetonitrile\n",
      "\t(PHC) L-1-pyrroline-3-hydroxy-5-carboxylate\n",
      "\t(LACT) lactose\n",
      "\t(PYTE) myo-inositol hexakisphosphate\n",
      "\t(MHIS) N(pi)-methyl-L-histidine\n",
      "\t(ACVb) N-[L-5-amino-5-carboxypentanoyl]-L-cysteinyl-D-valine\n",
      "\t(OSLHSER) O-succinyl-L-homoserine\n",
      "\t(PROPAL) propanal\n",
      "\t(PURI5P) pseudouridine 5'-phosphate\n",
      "\t(QT) quinate\n",
      "\t(SARC) sarcosine\n",
      "\t(SPRM) spermine\n",
      "\t(T6P) tagatose 6-phosphate\n",
      "\t(TAR) tartrate\n",
      "\t(THPTGLU) tetrahydropteroyltri-L-glutamate\n",
      "\t(DT) thymidine\n",
      "\t(THY) thymine\n",
      "\t(HPRO) trans-4-hydroxy-L-proline\n",
      "\t(TRPM) tryptamine\n",
      "\t(UDPGE) UDP-D-glucuronate\n",
      "\t(GALUNTe) D-galacturonate\n",
      "\t(PECTATEe) pectate\n",
      "\t(PHEETHALm) 2-phenylethanol\n",
      "\t(ECYSm) cysteine-[enzyme]\n",
      "\t(FALDm) formaldehyde\n",
      "\t(PHCm) L-1-pyrroline-3-hydroxy-5-carboxylate\n",
      "\t(ALAm) L-alanine\n",
      "\t(DIMEGLYm) N,N-dimethylglycine\n",
      "\t(NMNm) nicotinamide mononucleotide\n",
      "\t(OAHSERm) O-acetyl-L-homoserine\n",
      "\t(PHALm) phenylacetaldehyde\n",
      "\t(SARCm) sarcosine\n",
      "\t(ESULFCYSm) S-sulfanylcysteine-[enzyme]\n",
      "\t(HPROm) trans-4-hydroxy-L-proline\n",
      "\t(LYSp) L-lysine\n",
      "\n",
      "Reactions which could not be elementally balanced\t43\n",
      "\tr0123\n",
      "\tr0144\n",
      "\tr0244\n",
      "\tr0247\n",
      "\tr0248\n",
      "\tr0280\n",
      "\tr0281\n",
      "\tr0282\n",
      "\tr0283\n",
      "\tr0302\n",
      "\tr0303\n",
      "\tr0304\n",
      "\tr0312\n",
      "\tr0313\n",
      "\tr0481\n",
      "\tr0482\n",
      "\tr0514\n",
      "\tr1010\n",
      "\tr1109\n",
      "\tr1119\n",
      "\tr1120\n",
      "\tr1121\n",
      "\tr1125\n",
      "\tr1134\n",
      "\tr1135\n",
      "\tr1261\n",
      "\tr1348\n",
      "\tr1422\n",
      "\tr1453\n",
      "\tr1455\n",
      "\tr1456\n",
      "\tr1459\n",
      "\tr1460\n",
      "\tr1461\n",
      "\tr1462\n",
      "\tr1463\n",
      "\tr1464\n",
      "\tr1465\n",
      "\tr1467\n",
      "\tr1468\n",
      "\tpenartOUT\n",
      "\tproteinOUT\n",
      "\tbmOUT\n",
      "\n",
      "Reactions which are elementally unbalanced\t123\n",
      "\tr0065\thydrogen\n",
      "\tr0066\thydrogen\n",
      "\tr0108\thydrogen\n",
      "\tr0118\thydrogen\n",
      "\tr0121\thydrogen\n",
      "\tr0127\thydrogen\n",
      "\tr0134\thydrogen\n",
      "\tr0137\toxygen, hydrogen\n",
      "\tr0148\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0161\tgeneric group\n",
      "\tr0174\toxygen, hydrogen, generic group\n",
      "\tr0202\toxygen, hydrogen\n",
      "\tr0205\toxygen, hydrogen\n",
      "\tr0213\thydrogen\n",
      "\tr0223\toxygen, hydrogen\n",
      "\tr0261\toxygen, hydrogen\n",
      "\tr0264\toxygen, hydrogen\n",
      "\tr0277\toxygen, hydrogen\n",
      "\tr0278\toxygen, hydrogen\n",
      "\tr0284\toxygen, hydrogen\n",
      "\tr0285\tcarbon, oxygen, hydrogen\n",
      "\tr0286\toxygen, hydrogen\n",
      "\tr0287\toxygen, hydrogen\n",
      "\tr0288\toxygen, hydrogen\n",
      "\tr0289\tcarbon, oxygen, hydrogen\n",
      "\tr0298\thydrogen\n",
      "\tr0358\toxygen, hydrogen\n",
      "\tr0359\toxygen, hydrogen\n",
      "\tr0360\toxygen, hydrogen\n",
      "\tr0478\thydrogen, iron\n",
      "\tr0479\tcarbon, oxygen, hydrogen\n",
      "\tr0480\toxygen, hydrogen\n",
      "\tr0483\tiron\n",
      "\tr0484\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0486\thydrogen\n",
      "\tr0491\thydrogen\n",
      "\tr0506\tcarbon, oxygen, phosphorus, hydrogen\n",
      "\tr0507\tcarbon, oxygen, phosphorus, hydrogen\n",
      "\tr0519\thydrogen\n",
      "\tr0520\thydrogen\n",
      "\tr0521\toxygen, hydrogen\n",
      "\tr0533\thydrogen\n",
      "\tr0534\toxygen, hydrogen\n",
      "\tr0561\thydrogen\n",
      "\tr0564\thydrogen\n",
      "\tr0586\thydrogen\n",
      "\tr0629\toxygen, hydrogen\n",
      "\tr0654\thydrogen\n",
      "\tr0697\thydrogen\n",
      "\tr0729\tcarbon, oxygen, hydrogen\n",
      "\tr0745\toxygen\n",
      "\tr0769\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0771\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0772\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0773\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0774\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0775\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0781\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0782\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0783\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0784\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0786\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0787\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0788\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0789\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0790\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0792\tcarbon, nitrogen, hydrogen, generic group\n",
      "\tr0822\thydrogen\n",
      "\tr0833\thydrogen\n",
      "\tr1003\thydrogen\n",
      "\tr1004\thydrogen\n",
      "\tr1005\thydrogen\n",
      "\tr1006\thydrogen\n",
      "\tr1007\thydrogen\n",
      "\tr1008\thydrogen\n",
      "\tr1009\thydrogen\n",
      "\tr1011\thydrogen\n",
      "\tr1012\thydrogen\n",
      "\tr1013\thydrogen\n",
      "\tr1014\thydrogen\n",
      "\tr1015\thydrogen\n",
      "\tr1020\thydrogen\n",
      "\tr1029\thydrogen\n",
      "\tr1030\thydrogen\n",
      "\tr1035\thydrogen\n",
      "\tr1036\thydrogen\n",
      "\tr1041\thydrogen\n",
      "\tr1042\thydrogen\n",
      "\tr1046\thydrogen\n",
      "\tr1047\thydrogen\n",
      "\tr1048\thydrogen\n",
      "\tr1049\thydrogen\n",
      "\tr1051\thydrogen\n",
      "\tr1052\thydrogen\n",
      "\tr1056\thydrogen\n",
      "\tr1057\thydrogen\n",
      "\tr1061\thydrogen\n",
      "\tr1062\thydrogen\n",
      "\tr1077\thydrogen\n",
      "\tr1078\thydrogen\n",
      "\tr1082\thydrogen\n",
      "\tr1083\thydrogen\n",
      "\tr1096\thydrogen\n",
      "\tr1097\thydrogen\n",
      "\tr1098\thydrogen\n",
      "\tr1099\thydrogen\n",
      "\tr1104\thydrogen\n",
      "\tr1106\thydrogen\n",
      "\tr1123\thydrogen\n",
      "\tr1127\thydrogen\n",
      "\tr1130\thydrogen\n",
      "\tr1131\thydrogen\n",
      "\tr1146\tcarbon, hydrogen\n",
      "\tr1152\thydrogen\n",
      "\tr1153\thydrogen\n",
      "\tr1158\thydrogen\n",
      "\tr1159\thydrogen\n",
      "\tr1161\thydrogen\n",
      "\tr1454\tcarbon, hydrogen, generic group\n",
      "\tr1457\tcarbon, nitrogen, oxygen, phosphorus, hydrogen, generic group\n",
      "\tr1458\tcarbon, nitrogen, oxygen, phosphorus, hydrogen, generic group\n",
      "\tfreeNADH\thydrogen\n",
      "\tfreeNADPH\thydrogen\n",
      "\n"
     ]
    }
   ],
   "source": [
    "% Usefull for model validation. Notice the two \"true\" flags.\n",
    "printModelStats(model,true,true);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the tutorial:\n",
    "\n",
    "```MATLAB\n",
    "%Most modelling approaches using GEMs are based on the mass balancing\n",
    "%around the internal metabolites in the system. However, in order for the\n",
    "%system to uptake or excrete metabolites, some metabolites have been\n",
    "%defined as \"unconstrained\". In order to simulate something, those\n",
    "%metabolites have to be removed from the model. The function simplifyModel\n",
    "%is a general-purpose function for making models smaller. This includes the\n",
    "%options such as grouping linear reactions and deleting reactions which\n",
    "%cannot carry flux. Here it is chosen to delete the exchange metabolites,\n",
    "%all reactions that are constrained to zero (mainly uptake of non-standard\n",
    "%carbon sources), and all reactions that cannot carry flux (mainly\n",
    "%reactions that were dependent on any of those non-standard carbons\n",
    "%sources).\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [simplifyModel](https://sysbiochalmers.github.io/RAVEN/doc/core/simplifyModel.html)\n",
    "\n",
    "```MATLAB\n",
    "function [reducedModel, deletedReactions, deletedMetabolites]=simplifyModel(model,deleteUnconstrained, \n",
    "deleteDuplicates, deleteZeroInterval, deleteInaccessible, deleteMinMax, groupLinear, constrainReversible, reservedRxns, suppressWarnings)\n",
    "```\n",
    "\n",
    "Simplifies a model by deleting reactions/metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<html><body><pre>model = struct with fields:\n",
       "   id: 'iAL1006'\n",
       "              name: 'Penicillium chrysogenum genome-scale model'\n",
       "        annotation: [1x1 struct]\n",
       "              rxns: {1305x1 cell}\n",
       "              mets: {1037x1 cell}\n",
       "                 S: [1037x1305 double]\n",
       "                lb: [1305x1 double]\n",
       "                ub: [1305x1 double]\n",
       "               rev: [1305x1 double]\n",
       "                 c: [1305x1 double]\n",
       "                 b: [1037x1 double]\n",
       "             comps: {4x1 cell}\n",
       "         compNames: {4x1 cell}\n",
       "       compMiriams: {4x1 cell}\n",
       "          rxnNames: {1305x1 cell}\n",
       "          rxnComps: [1305x1 double]\n",
       "           grRules: {1305x1 cell}\n",
       "        rxnGeneMat: [1305x1006 double]\n",
       "        subSystems: {1305x1 cell}\n",
       "           eccodes: {1305x1 cell}\n",
       "        rxnMiriams: {1305x1 cell}\n",
       "          rxnNotes: {1305x1 cell}\n",
       "     rxnReferences: {1305x1 cell}\n",
       "             genes: {1006x1 cell}\n",
       "       geneMiriams: {1006x1 cell}\n",
       "    geneShortNames: {1006x1 cell}\n",
       "          metNames: {1037x1 cell}\n",
       "          metComps: [1037x1 double]\n",
       "            inchis: {1037x1 cell}\n",
       "       metFormulas: {1037x1 cell}\n",
       "        metMiriams: {1037x1 cell}</pre></body></html>"
      ],
      "text/plain": [
       "model = struct with fields:\n",
       "   id: 'iAL1006'\n",
       "              name: 'Penicillium chrysogenum genome-scale model'\n",
       "        annotation: [1x1 struct]\n",
       "              rxns: {1305x1 cell}\n",
       "              mets: {1037x1 cell}\n",
       "                 S: [1037x1305 double]\n",
       "                lb: [1305x1 double]\n",
       "                ub: [1305x1 double]\n",
       "               rev: [1305x1 double]\n",
       "                 c: [1305x1 double]\n",
       "                 b: [1037x1 double]\n",
       "             comps: {4x1 cell}\n",
       "         compNames: {4x1 cell}\n",
       "       compMiriams: {4x1 cell}\n",
       "          rxnNames: {1305x1 cell}\n",
       "          rxnComps: [1305x1 double]\n",
       "           grRules: {1305x1 cell}\n",
       "        rxnGeneMat: [1305x1006 double]\n",
       "        subSystems: {1305x1 cell}\n",
       "           eccodes: {1305x1 cell}\n",
       "        rxnMiriams: {1305x1 cell}\n",
       "          rxnNotes: {1305x1 cell}\n",
       "     rxnReferences: {1305x1 cell}\n",
       "             genes: {1006x1 cell}\n",
       "       geneMiriams: {1006x1 cell}\n",
       "    geneShortNames: {1006x1 cell}\n",
       "          metNames: {1037x1 cell}\n",
       "          metComps: [1037x1 double]\n",
       "            inchis: {1037x1 cell}\n",
       "       metFormulas: {1037x1 cell}\n",
       "        metMiriams: {1037x1 cell}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model=simplifyModel(model,true,false,true,true);\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [setParam](https://sysbiochalmers.github.io/RAVEN/doc/core/setParam.html)\n",
    "\n",
    "\n",
    "```MATLAB\n",
    "function model=setParam(model, paramType, rxnList, params, var)\n",
    "```\n",
    "\n",
    "Sets parameters for reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%%% validating the model using the theoretical yield of carbon dioxide from glucose. %%%%\n",
    "% First, set the uptake of carbon to glucose\n",
    "model=setParam(model,'ub',{'glcIN' 'etohIN'},[1 0]); % remember, the units are mmol/gDW/h\n",
    "% Second, set the objective for the simulation to maximize CO2 production\n",
    "model=setParam(model,'obj',{'co2OUT'},1);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [solveLP](https://sysbiochalmers.github.io/RAVEN/doc/solver/solveLP.html)\n",
    "\n",
    "``` MATLAB\n",
    "function [solution, hsSolOut]=solveLP(model,minFlux,params,hsSol)\n",
    "```\n",
    "\n",
    "Solves a linear programming problem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         x: [1305x1 double]\n",
      "         f: -6.0000\n",
      "      stat: 1\n",
      "       msg: 'Optimal solution found'\n",
      "    sPrice: [1037x1 double]\n",
      "     rCost: [1305x1 double]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "% Third, get the maximum given the setted constrains\n",
    "sol=solveLP(model);\n",
    "disp(sol);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [printFluxes](https://sysbiochalmers.github.io/RAVEN/doc/core/printFluxes.html)\n",
    "\n",
    "```MATLAB\n",
    "function printFluxes(model, fluxes, onlyExchange, cutOffFlux, outputFile,outputString,metaboliteList)\n",
    "```\n",
    "\n",
    "Prints reactions and fluxes to the screen or to a file.\n",
    "\n",
    "> onlyExchange: only print exchange fluxes (opt, default true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EXCHANGE FLUXES:\n",
      "co2OUT\t(production of CO2):\t6\n",
      "h2oOUT\t(production of H2O):\t6\n",
      "glcIN\t(uptake of alpha-D-glucose):\t1\n",
      "o2IN\t(uptake of O2):\t6\n"
     ]
    }
   ],
   "source": [
    "% Print the fluxes in console. The true flag indicates to only print exchange fluxes\n",
    "printFluxes(model, sol.x, true, 10^-7);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FLUXES:\n",
      "r0001\t(spontaneous conversion):\t333.3333\n",
      "r0002\t(spontaneous conversion):\t-333.3333\n",
      "r0004\t(ATP:alpha-D-glucose 6-phosphotransferase):\t1\n",
      "r0006\t(alpha-D-glucose 6-phosphate ketol-isomerase):\t-999\n",
      "r0007\t(beta-D-glucose 6-phosphate ketol-isomerase):\t-1000\n",
      "r0008\t(alpha-D-glucose 6-phosphate ketol-isomerase):\t1000\n",
      "r0009\t(phosphofructokinase):\t1\n",
      "r0010\t(fructose-bisphosphate aldolase):\t1\n",
      "r0011\t(D-glyceraldehyde-3-phosphate aldose-ketose-isomerase):\t-1\n",
      "r0012\t(glyceraldehyde-3-phosphate dehydrogenase):\t2\n",
      "r0013\t(phosphoglycerate kinase):\t2\n",
      "r0014\t(phosphoglycerate mutase):\t2\n",
      "r0015\t(2-phospho-D-glycerate hydro-lyase (phosphoenolpyruvate-forming)):\t2\n",
      "r0016\t(pyruvate kinase):\t2\n",
      "r0019\t(ethanol:NADP+ oxidoreductase):\t8\n",
      "r0020\t(ethanol:NAD+ oxidoreductase):\t-1000\n",
      "r0021\t(ethanol:NAD+ oxidoreductase):\t992\n",
      "r0022\t(pyruvate:[dihydrolipoyllysine-residue acetyltransferase]-lipoyllysine 2-oxidoreductase (decarboxylating, acceptor-acetylating)):\t2\n",
      "r0023\t(acetyl-CoA:enzyme N6-(dihydrolipoyl)lysine S-acetyltransferase):\t2\n",
      "r0024\t(dihydrolipoamide:NAD+ oxidoreductase):\t4\n",
      "r0081\t((S)-3-hydroxybutanoyl-CoA:NADP+ oxidoreductase):\t-2\n",
      "r0084\t(succinate:ubiquinone oxidoreductase):\t-2\n",
      "r0085\t(acetyl-CoA:oxaloacetate C-acetyltransferase (thioester-hydrolysing)):\t2\n",
      "r0086\t(citrate hydro-lyase (cis-aconitate-forming)):\t-2\n",
      "r0087\t(isocitrate hydro-lyase (cis-aconitate-forming)):\t2\n",
      "r0089\t(isocitrate:NADP+ oxidoreductase (decarboxylating)):\t2\n",
      "r0091\t(2-oxoglutarate:[dihydrolipoyllysine-residue succinyltransferase]-lipoyllysine 2-oxidoreductase (decarboxylating, acceptor-succinylating)):\t2\n",
      "r0092\t(succinyl-CoA:enzyme N6-(dihydrolipoyl)lysine S-succinyltransferase):\t2\n",
      "r0094\t(succinate:CoA ligase (GDP-forming)):\t-998\n",
      "r0095\t(succinate:CoA ligase (ADP-forming)):\t1000\n",
      "r0097\t((S)-malate hydro-lyase (fumarate-forming)):\t2\n",
      "r0098\t((S)-malate:NAD+ oxidoreductase):\t-998\n",
      "r0099\t((S)-malate:NAD+ oxidoreductase):\t1000\n",
      "r0116\t(ubiquinol oxidase (mitochondrial alternative oxidase (aox))):\t5.6646\n",
      "r0118\t(ferrocytochrome-c:oxygen oxidoreductase):\t0.1677\n",
      "r0120\t(NADH:ubiquinone oxidoreductase):\t4\n",
      "r0127\t(ubiquinol:ferricytochrome-c oxidoreductase):\t0.3354\n",
      "r0128\t(H+-transporting two-sector ATPase):\t0.14\n",
      "r0149\t(5,10-methylenetetrahydrofolate:glycine hydroxymethyltransferase):\t-1000\n",
      "r0150\t(5,10-methylenetetrahydrofolate:glycine hydroxymethyltransferase):\t1000\n",
      "r0175\t(D-mannitol:NADP+ 2-oxidoreductase):\t8\n",
      "r0181\t(D-mannitol:NAD+ 2-oxidoreductase):\t-8\n",
      "r0205\t(cellobiose dehydrogenase):\t-994\n",
      "r0223\t(cellobiose dehydrogenase):\t994\n",
      "r0226\t(UTP:alpha-D-galactose-1-phosphate uridylyltransferase):\t1000\n",
      "r0228\t(UTP:alpha-D-glucose-1-phosphate uridylyltransferase):\t-1000\n",
      "r0230\t(UDP-glucose:alpha-D-galactose-1-phosphate uridylyltransferase):\t-1000\n",
      "r0233\t(D-galactose:oxygen 6-oxidoreductase):\t6\n",
      "r0236\t(D-galactose:NAD+ 1-oxidoreductase):\t-6\n",
      "r0237\t(D-galactono-1,4-lactone hydroxyacylhydrolase):\t-6\n",
      "r0374\t(ATP:AMP phosphotransferase):\t-1000\n",
      "r0376\t(adenylate kinase):\t1000\n",
      "r0377\t(adenylate kinase):\t-998\n",
      "r0378\t(adenylate kinase):\t998\n",
      "r0398\t(ATP:dADP phosphotransferase):\t1000\n",
      "r0399\t(ATP:GDP phosphotransferase):\t-1000\n",
      "r0400\t(ATP:IDP phosphotransferase):\t-1000\n",
      "r0401\t(ATP:GMP phosphotransferase):\t-1000\n",
      "r0403\t(guanylate kinase):\t-1000\n",
      "r0416\t(GTP:uridine 5'-phosphotransferase):\t1000\n",
      "r0418\t(ATP:uridine 5'-phosphotransferase):\t1000\n",
      "r0540\t(L-aspartate:2-oxoglutarate aminotransferase):\t1000\n",
      "r0541\t(L-aspartate:2-oxoglutarate aminotransferase):\t-1000\n",
      "r0735\t(hydrogen-peroxide:hydrogen-peroxide oxidoreductase):\t3\n",
      "r0929\t(3-hydroxyacyl-CoA dehydrogenase):\t2\n",
      "r1165\t((S)-malate mitochondrial permease):\t1000\n",
      "r1170\t(2-oxoglutarate mitochondrial permease):\t1000\n",
      "r1174\t(5,10-methylenetetrahydrofolate mitochondrial permease):\t-1000\n",
      "r1178\t(ADP/ATP/phosphate mitochondrial shuttle):\t2.14\n",
      "r1182\t(citrate/isocitrate mitochondrial shuttle):\t-1000\n",
      "r1183\t(citrate/malate mitochondrial shuttle):\t-1000\n",
      "r1194\t(glycine mitochondrial permease):\t-1000\n",
      "r1200\t(isocitrate/malate mitochondrial shuttle):\t1000\n",
      "r1204\t(L-aspartate mitochondrial permease):\t1000\n",
      "r1217\t(malate/succinate mitochondrial shuttle):\t-1000\n",
      "r1230\t(phosphate mitochondrial permease):\t-1000\n",
      "r1231\t(phosphate/malate mitochondrial shuttle):\t-1000\n",
      "r1235\t(pyruvate mitochondrial permease):\t2\n",
      "r1236\t(serine mitochondrial permease):\t1000\n",
      "r1238\t(succinate mitochondrial permease):\t-1000\n",
      "r1240\t(tetrahydrofolate mitochondrial permease):\t1000\n",
      "r1243\t(acetaldehyde mitochondrial membrane diffusion):\t-992\n",
      "r1246\t(CO2 mitochondrial membrane diffusion):\t-6\n",
      "r1247\t(ethanol mitochondrial membrane diffusion):\t992\n",
      "r1248\t(H2O mitochondrial membrane diffusion):\t1000\n",
      "r1251\t(O2 mitochondrial membrane diffusion):\t3\n",
      "r1316\t(alpha-D-glucose permease):\t334.3333\n",
      "r1322\t(beta-D-glucose permease):\t666.6667\n",
      "r1323\t(cellobiono-1,5-lactone permease):\t-994\n",
      "r1324\t(cellobiose permease):\t994\n",
      "r1336\t(D-mannose permease):\t1000\n",
      "r1337\t(D-mannose permease):\t-1000\n",
      "r1342\t(glucose permease):\t-1000\n",
      "r1424\t(CO2 plasma membrane diffusion):\t6\n",
      "r1428\t(H2O plasma membrane diffusion):\t6\n",
      "r1441\t(O2 plasma membrane diffusion):\t1000\n",
      "r1466\t(ATP maintenance):\t1\n",
      "co2OUT\t(production of CO2):\t6\n",
      "h2oOUT\t(production of H2O):\t6\n",
      "glcIN\t(uptake of alpha-D-glucose):\t1\n",
      "o2IN\t(uptake of O2):\t6\n"
     ]
    }
   ],
   "source": [
    "% Exploring other fluxes. Notice that some are \"infinite\" (1000)\n",
    "printFluxes(model, sol.x, false, 10^-7);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\"Infinite\" (±1000) fluxes suggest and non physiologicla behaviour. The tutorial states:\n",
    "\n",
    "```MATLAB\n",
    "%The results show many reactions that have -1000 or 1000 flux. This is\n",
    "%because there are loops in the solution. In order to clean up the solution\n",
    "%one can minimize the sum of all the fluxes. This is done by setting the\n",
    "%third flag to solveLP to true (take a look at solveLP, there are other\n",
    "%options as well).\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FLUXES:\n",
      "r0003\t(D-glucose 1-epimerase):\t1\n",
      "r0006\t(alpha-D-glucose 6-phosphate ketol-isomerase):\t-2.5446\n",
      "r0011\t(D-glyceraldehyde-3-phosphate aldose-ketose-isomerase):\t0.37625\n",
      "r0012\t(glyceraldehyde-3-phosphate dehydrogenase):\t0.41584\n",
      "r0013\t(phosphoglycerate kinase):\t0.41584\n",
      "r0014\t(phosphoglycerate mutase):\t0.41584\n",
      "r0015\t(2-phospho-D-glycerate hydro-lyase (phosphoenolpyruvate-forming)):\t0.41584\n",
      "r0022\t(pyruvate:[dihydrolipoyllysine-residue acetyltransferase]-lipoyllysine 2-oxidoreductase (decarboxylating, acceptor-acetylating)):\t0.20792\n",
      "r0023\t(acetyl-CoA:enzyme N6-(dihydrolipoyl)lysine S-acetyltransferase):\t0.20792\n",
      "r0024\t(dihydrolipoamide:NAD+ oxidoreductase):\t0.83167\n",
      "r0027\t(glucose-6-phosphate 1-dehydrogenase):\t2.5446\n",
      "r0028\t(6-phospho-D-glucono-1,5-lactone lactonohydrolase):\t2.5446\n",
      "r0029\t(6-phospho-D-gluconate:NADP+ 2-oxidoreductase (decarboxylating)):\t3.5446\n",
      "r0030\t(D-ribulose-5-phosphate 3-epimerase):\t2.1683\n",
      "r0031\t(D-ribose-5-phosphate aldose-ketose-isomerase):\t1.3762\n",
      "r0032\t(sedoheptulose-7-phosphate:D-glyceraldehyde-3-phosphate glycolaldehyde transferase):\t1.3762\n",
      "r0033\t(transketolase):\t1.1683\n",
      "r0034\t(sedoheptulose-7-phosphate:D-glyceraldehyde-3-phosphate glyceronetransferase):\t-1.3762\n",
      "r0038\t(ATP:D-gluconate 6-phosphotransferase):\t1\n",
      "r0042\t(beta-D-glucose:oxygen 1-oxidoreductase):\t-1\n",
      "r0045\t(spontaneous conversion):\t1\n",
      "r0059\t((S)-malate:NADP+ oxidoreductase(oxaloacetate-decarboxylating)):\t0.20792\n",
      "r0084\t(succinate:ubiquinone oxidoreductase):\t-0.62375\n",
      "r0085\t(acetyl-CoA:oxaloacetate C-acetyltransferase (thioester-hydrolysing)):\t0.62375\n",
      "r0086\t(citrate hydro-lyase (cis-aconitate-forming)):\t-0.62375\n",
      "r0087\t(isocitrate hydro-lyase (cis-aconitate-forming)):\t0.62375\n",
      "r0090\t(isocitrate:NADP+ oxidoreductase (decarboxylating)):\t0.62375\n",
      "r0091\t(2-oxoglutarate:[dihydrolipoyllysine-residue succinyltransferase]-lipoyllysine 2-oxidoreductase (decarboxylating, acceptor-succinylating)):\t0.62375\n",
      "r0092\t(succinyl-CoA:enzyme N6-(dihydrolipoyl)lysine S-succinyltransferase):\t0.62375\n",
      "r0093\t(succinyl-CoA:acetoacetate CoA-transferase):\t0.20792\n",
      "r0095\t(succinate:CoA ligase (ADP-forming)):\t0.41584\n",
      "r0097\t((S)-malate hydro-lyase (fumarate-forming)):\t0.83167\n",
      "r0098\t((S)-malate:NAD+ oxidoreductase):\t0.62375\n",
      "r0118\t(ferrocytochrome-c:oxygen oxidoreductase):\t1.0396\n",
      "r0119\t(NADH:ubiquinone oxidoreductase):\t1.4554\n",
      "r0124\t(NADPH:oxygen oxidoreductase):\t6.2971\n",
      "r0127\t(ubiquinol:ferricytochrome-c oxidoreductase):\t2.0792\n",
      "r0128\t(H+-transporting two-sector ATPase):\t4.5162\n",
      "r0153\t(formaldehyde:NAD+ oxidoreductase):\t0.37625\n",
      "r0157\t(formate:NAD+ oxidoreductase):\t0.37625\n",
      "r0158\t(D-xylulose-5-phosphate:formaldehyde glycolaldehydetransferase):\t0.37625\n",
      "r0242\t(sn-glycerol-3-phosphate phosphohydrolase):\t0.37625\n",
      "r0253\t(alcohol oxidase):\t0.58416\n",
      "r0254\t(alcohol oxidase):\t0.37625\n",
      "r0255\t(NAD-dependent alcohol dehydrogenase):\t0.58416\n",
      "r0308\t(sn-glycerol-3-phosphate:NAD+ 2-oxidoreductase):\t0.37625\n",
      "r0553\t(L-glutamate:NADP+ oxidoreductase (deaminating)):\t0.20792\n",
      "r0693\t(phosphoenolpyruvate:D-erythrose-4-phosphate C-(1-carboxyvinyl)transferase (phosphate hydrolysing, 2-carboxy-2-oxoethyl-forming)):\t0.20792\n",
      "r0694\t( 2-dehydro-3-deoxy-D-arabino-heptonate 7-phosphate phosphate-lyase (cyclyzing)):\t0.20792\n",
      "r0695\t(3-dehydroquinate hydro-lyase):\t0.20792\n",
      "r0696\t(shikimate:NADP+ 3-oxidoreductase):\t0.20792\n",
      "r0698\t(ATP:shikimate 3-phosphotransferase):\t0.20792\n",
      "r0699\t(phosphoenolpyruvate:3-phosphoshikimate 5-O-(1-carboxyvinyl)-transferase):\t0.20792\n",
      "r0700\t(5-O-(1-carboxyvinyl)-3-phosphoshikimate phosphate-lyase (chorismate-forming)):\t0.20792\n",
      "r0713\t(chorismate pyruvatemutase):\t0.20792\n",
      "r0716\t(prephenate hydro-lyase (decarboxylating; phenylpyruvate-forming)):\t0.20792\n",
      "r0717\t(L-phenylalanine:2-oxoglutarate aminotransferase):\t-0.20792\n",
      "r0720\t(homogentisate:oxygen 1,2-oxidoreductase (decyclizing)):\t0.20792\n",
      "r0721\t(maleylacetoacetate isomerase):\t0.20792\n",
      "r0722\t(4-fumarylacetoacetate fumarylhydrolase):\t0.20792\n",
      "r0724\t(L-phenylalanine:oxygen oxidoreductase (decarboxylating)):\t0.20792\n",
      "r0725\t(2-phenylacetamide amidohydrolase):\t0.20792\n",
      "r0735\t(hydrogen-peroxide:hydrogen-peroxide oxidoreductase):\t4.1287\n",
      "r0748\t(phenylacetate 2-hydroxylase):\t0.20792\n",
      "r0751\t(2-hydroxyphenylacetate hydroxylase):\t0.20792\n",
      "r0930\t(acetyl-CoA:acetyl-CoA C-acetyltransferase):\t0.20792\n",
      "r1164\t((S)-malate mitochondrial permease):\t-0.20792\n",
      "r1170\t(2-oxoglutarate mitochondrial permease):\t0.41584\n",
      "r1178\t(ADP/ATP/phosphate mitochondrial shuttle):\t4.9321\n",
      "r1191\t(fumarate mitochondrial permease):\t0.20792\n",
      "r1199\t(isocitrate mitochondrial permease):\t0.62375\n",
      "r1235\t(pyruvate mitochondrial permease):\t0.20792\n",
      "r1245\t(acetoacetate mitochondrial membrane diffusion):\t0.20792\n",
      "r1246\t(CO2 mitochondrial membrane diffusion):\t-0.83167\n",
      "r1248\t(H2O mitochondrial membrane diffusion):\t0.20792\n",
      "r1251\t(O2 mitochondrial membrane diffusion):\t1.0396\n",
      "r1316\t(alpha-D-glucose permease):\t1\n",
      "r1424\t(CO2 plasma membrane diffusion):\t6\n",
      "r1428\t(H2O plasma membrane diffusion):\t6\n",
      "r1441\t(O2 plasma membrane diffusion):\t6\n",
      "r1466\t(ATP maintenance):\t1\n",
      "co2OUT\t(production of CO2):\t6\n",
      "h2oOUT\t(production of H2O):\t6\n",
      "glcIN\t(uptake of alpha-D-glucose):\t1\n",
      "o2IN\t(uptake of O2):\t6\n"
     ]
    }
   ],
   "source": [
    "% minFLux = 1 indicates that a second optimization performed to get rid of loops in the flux distribution will\n",
    "% adress to minimize the sum of abs(fluxes). \n",
    "sol=solveLP(model,1);\n",
    "printFluxes(model, sol.x, false, 10^-7);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EXCHANGE FLUXES:\n",
      "c4odOUT\t(production of 2-oxy-but-3-enoate):\t8.4803e-06\n",
      "bmOUT\t(production of biomass):\t0.084803\n",
      "co2OUT\t(production of CO2):\t3.061\n",
      "h2oOUT\t(production of H2O):\t5.586\n",
      "glcIN\t(uptake of alpha-D-glucose):\t1\n",
      "piIN\t(uptake of phosphate):\t0.027889\n",
      "nh3IN\t(uptake of NH3):\t0.59384\n",
      "o2IN\t(uptake of O2):\t2.937\n",
      "slfIN\t(uptake of sulfate):\t0.022888\n",
      "thmIN\t(uptake of thiamin):\t8.4803e-06\n",
      "pimIN\t(uptake of pimelate):\t8.4803e-06\n"
     ]
    }
   ],
   "source": [
    "% Lets repeat, but change the objective function to biomass production\n",
    "% Notice the change in metabolic requirements (exchange IN reactions)\n",
    "model=setParam(model,'obj',{'bmOUT'},1);\n",
    "sol=solveLP(model,1);\n",
    "printFluxes(model, sol.x, true, 10^-7);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EXCHANGE FLUXES:\n",
      "c4odOUT\t(production of 2-oxy-but-3-enoate):\t1.0816e-05\n",
      "bmOUT\t(production of biomass):\t0.10816\n",
      "co2OUT\t(production of CO2):\t2.2514\n",
      "h2oOUT\t(production of H2O):\t8.4719\n",
      "piIN\t(uptake of phosphate):\t0.035572\n",
      "nh3IN\t(uptake of NH3):\t0.75743\n",
      "o2IN\t(uptake of O2):\t5.0932\n",
      "slfIN\t(uptake of sulfate):\t0.029193\n",
      "thmIN\t(uptake of thiamin):\t1.0816e-05\n",
      "pimIN\t(uptake of pimelate):\t1.0816e-05\n",
      "etohIN\t(uptake of ethanol):\t3\n"
     ]
    }
   ],
   "source": [
    "% To compare carbon flux swapping glucose by ethanol, remembear that\n",
    "% ethanol have 2 carbons and glucose 6; so, 3 * 2 = carbon number in glucose\n",
    "modelETH=setParam(model,'eq',{'glcIN' 'etohIN'},[0 3]);\n",
    "solETH=solveLP(modelETH,1);\n",
    "printFluxes(modelETH, solETH.x, true, 10^-7);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To make more clear the comparisson between two conditions:\n",
    "\n",
    "- [followChanged](https://sysbiochalmers.github.io/RAVEN/doc/core/followChanged.html)\n",
    "\n",
    "```MATLAB\n",
    "function followChanged(model,fluxesA,fluxesB, cutOffChange, cutOffFlux, cutOffDiff, metaboliteList)\n",
    "```\n",
    "\n",
    "Prints fluxes and reactions for each of the reactions that results in different fluxes compared to the reference case (fluxesB).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These reactions have flux values that differ by more than 50 percent, absolute values above 0.5, and a total difference above 0.5 (64 reactions)\n",
      "\n",
      "r0004: alpha-D-glucose[c] + ATP[c] => ADP[c] + alpha-D-glucose 6-phosphate[c]\n",
      "\tATP:alpha-D-glucose 6-phosphotransferase\n",
      "\tFlux: 1 Reference flux: 0 Difference: 1\n",
      "\n",
      "r0006: alpha-D-glucose 6-phosphate[c] <=> beta-D-fructofuranose 6-phosphate[c]\n",
      "\talpha-D-glucose 6-phosphate ketol-isomerase\n",
      "\tFlux: 0.91584 Reference flux: -0.10723 Difference: 1.0231\n",
      "\n",
      "r0009: ATP[c] + beta-D-fructofuranose 6-phosphate[c] => ADP[c] + beta-D-fructofuranose 1,6-bisphosphate[c]\n",
      "\tphosphofructokinase\n",
      "\tFlux: 0.87408 Reference flux: 0 Difference: 0.87408\n",
      "\n",
      "r0010: beta-D-fructofuranose 1,6-bisphosphate[c] <=> D-glyceraldehyde 3-phosphate[c] + glycerone phosphate[c]\n",
      "\tfructose-bisphosphate aldolase\n",
      "\tFlux: 0.87408 Reference flux: -0.16057 Difference: 1.0347\n",
      "\n",
      "r0011: D-glyceraldehyde 3-phosphate[c] <=> glycerone phosphate[c]\n",
      "\tD-glyceraldehyde-3-phosphate aldose-ketose-isomerase\n",
      "\tFlux: -0.82751 Reference flux: 0.21997 Difference: -1.0475\n",
      "\n",
      "r0012: D-glyceraldehyde 3-phosphate[c] + NAD(+)[c] + phosphate[c] <=> 3-phospho-D-glyceroyl phosphate[c] + NADH[c]\n",
      "\tglyceraldehyde-3-phosphate dehydrogenase\n",
      "\tFlux: 1.6507 Reference flux: -0.44547 Difference: 2.0962\n",
      "\n",
      "r0013: 3-phospho-D-glyceroyl phosphate[c] + ADP[c] <=> 3-phospho-D-glycerate[c] + ATP[c]\n",
      "\tphosphoglycerate kinase\n",
      "\tFlux: 1.6507 Reference flux: -0.44547 Difference: 2.0962\n",
      "\n",
      "r0014: 3-phospho-D-glycerate[c] <=> 2-phospho-D-glycerate[c]\n",
      "\tphosphoglycerate mutase\n",
      "\tFlux: 1.608 Reference flux: -0.44565 Difference: 2.0536\n",
      "\n",
      "r0015: 2-phospho-D-glycerate[c] <=> H2O[c] + phosphoenolpyruvate[c]\n",
      "\t2-phospho-D-glycerate hydro-lyase (phosphoenolpyruvate-forming)\n",
      "\tFlux: 1.608 Reference flux: -0.44565 Difference: 2.0536\n",
      "\n",
      "r0016: ADP[c] + phosphoenolpyruvate[c] => ATP[c] + pyruvate[c]\n",
      "\tpyruvate kinase\n",
      "\tFlux: 1.567 Reference flux: 0 Difference: 1.567\n",
      "\n",
      "r0019: ethanol[c] + NADP(+)[c] <=> acetaldehyde[c] + NADPH[c]\n",
      "\tethanol:NADP+ oxidoreductase\n",
      "\tFlux: 0 Reference flux: 1.6348 Difference: -1.6348\n",
      "\n",
      "r0020: ethanol[c] + NAD(+)[c] <=> acetaldehyde[c] + NADH[c]\n",
      "\tethanol:NAD+ oxidoreductase\n",
      "\tFlux: 0 Reference flux: 1.3652 Difference: -1.3652\n",
      "\n",
      "r0022: lipoamide-[enzyme][m] + pyruvate[m] => CO2[m] + S-acetyldihydrolipoamide-[enzyme][m]\n",
      "\tpyruvate:[dihydrolipoyllysine-residue acetyltransferase]-lipoyllysine 2-oxidoreductase (decarboxylating, acceptor-acetylating)\n",
      "\tFlux: 1.2352 Reference flux: 0 Difference: 1.2352\n",
      "\n",
      "r0023: coenzyme A[m] + S-acetyldihydrolipoamide-[enzyme][m] => acetyl-CoA[m] + dihydrolipoamide-[enzyme][m]\n",
      "\tacetyl-CoA:enzyme N6-(dihydrolipoyl)lysine S-acetyltransferase\n",
      "\tFlux: 1.2352 Reference flux: 0 Difference: 1.2352\n",
      "\n",
      "r0026: acetate[m] + ATP[m] + coenzyme A[m] => acetyl-CoA[m] + AMP[m] + diphosphate[m]\n",
      "\tacetate:CoA ligase (AMP-forming)\n",
      "\tFlux: 0 Reference flux: 1.8845 Difference: -1.8845\n",
      "\n",
      "r0053: acetaldehyde[m] + H2O[m] + NAD(+)[m] => acetate[m] + NADH[m]\n",
      "\tacetaldehyde:NAD+ oxidoreductase\n",
      "\tFlux: 0 Reference flux: 1.8845 Difference: -1.8845\n",
      "\n",
      "r0072: 2-oxobutanoate[m] + lipoamide-[enzyme][m] => CO2[m] + S-propionyldihydrolipoamide-[enzyme][m]\n",
      "\t2-oxobutanoate:[dihydrolipoyllysine-residue propionyltransferase]-lipoyllysine 2-oxidoreductase (decarboxylating, acceptor-propionylating)\n",
      "\tFlux: 0 Reference flux: 0.88113 Difference: -0.88113\n",
      "\n",
      "r0073: coenzyme A[m] + S-propionyldihydrolipoamide-[enzyme][m] => dihydrolipoamide-[enzyme][m] + propionyl-CoA[m]\n",
      "\tpropionyl-CoA:enzyme N6-(dihydrolipoyl)lysine S-propionyltransferase\n",
      "\tFlux: 0 Reference flux: 0.88113 Difference: -0.88113\n",
      "\n",
      "r0074: H2O[m] + oxaloacetate[m] + propionyl-CoA[m] => 2-methylcitrate[m] + coenzyme A[m]\n",
      "\tpropanoyl-CoA:oxaloacetate C-propanoyltransferase (thioester-hydrolysing, 1-carboxyethyl-forming)\n",
      "\tFlux: 0 Reference flux: 0.88098 Difference: -0.88098\n",
      "\n",
      "r0075: (Z)-but-2-ene-1,2,3-tricarboxylate[m] + H2O[m] <=> 2-methylcitrate[m]\n",
      "\t2-hydroxybutane-1,2,3-tricarboxylate hydro-lyase\n",
      "\tFlux: 0 Reference flux: -0.88098 Difference: 0.88098\n",
      "\n",
      "r0076: methylisocitrate[m] <=> pyruvate[m] + succinate[m]\n",
      "\t(2S,3R)-3-hydroxybutane-1,2,3-tricarboxylate pyruvate-lyase (succinate-forming)\n",
      "\tFlux: 0 Reference flux: 0.88098 Difference: -0.88098\n",
      "\n",
      "r0079: methylisocitrate[m] <=> (Z)-but-2-ene-1,2,3-tricarboxylate[m] + H2O[m]\n",
      "\t(2S,3R)-3-Hydroxybutane-1,2,3-tricarboxylate hydro-lyase\n",
      "\tFlux: 0 Reference flux: -0.88098 Difference: 0.88098\n",
      "\n",
      "r0084: fumarate[m] + ubiquinol[m] <=> succinate[m] + ubiquinone[m]\n",
      "\tsuccinate:ubiquinone oxidoreductase\n",
      "\tFlux: -0.91885 Reference flux: -2.5565 Difference: 1.6377\n",
      "\n",
      "r0085: acetyl-CoA[m] + H2O[m] + oxaloacetate[m] <=> citrate[m] + coenzyme A[m]\n",
      "\tacetyl-CoA:oxaloacetate C-acetyltransferase (thioester-hydrolysing)\n",
      "\tFlux: 1.1838 Reference flux: 1.8189 Difference: -0.6351\n",
      "\n",
      "r0086: cis-aconitate[m] + H2O[m] <=> citrate[m]\n",
      "\tcitrate hydro-lyase (cis-aconitate-forming)\n",
      "\tFlux: -1.0313 Reference flux: -1.8189 Difference: 0.78767\n",
      "\n",
      "r0087: cis-aconitate[m] + H2O[m] <=> isocitrate[m]\n",
      "\tisocitrate hydro-lyase (cis-aconitate-forming)\n",
      "\tFlux: 1.0313 Reference flux: 1.8189 Difference: -0.78767\n",
      "\n",
      "r0090: isocitrate[c] + NADP(+)[c] => 2-oxoglutarate[c] + CO2[c] + NADPH[c]\n",
      "\tisocitrate:NADP+ oxidoreductase (decarboxylating)\n",
      "\tFlux: 0.86372 Reference flux: 0.23939 Difference: 0.62433\n",
      "\n",
      "r0097: fumarate[m] + H2O[m] <=> (S)-malate[m]\n",
      "\t(S)-malate hydro-lyase (fumarate-forming)\n",
      "\tFlux: 0.96827 Reference flux: 2.6196 Difference: -1.6513\n",
      "\n",
      "r0098: (S)-malate[m] + NAD(+)[m] <=> NADH[m] + oxaloacetate[m]\n",
      "\t(S)-malate:NAD+ oxidoreductase\n",
      "\tFlux: 0.96827 Reference flux: 1.9921 Difference: -1.0239\n",
      "\n",
      "r0099: (S)-malate[c] + NAD(+)[c] <=> NADH[c] + oxaloacetate[c]\n",
      "\t(S)-malate:NAD+ oxidoreductase\n",
      "\tFlux: 0 Reference flux: 0.62742 Difference: -0.62742\n",
      "\n",
      "r0103: isocitrate[p] => glyoxylate[p] + succinate[p]\n",
      "\tisocitrate glyoxylate-lyase (succinate-forming)\n",
      "\tFlux: 0.055182 Reference flux: 1.0602 Difference: -1.005\n",
      "\n",
      "r0111: ATP[m] + CO2[m] + H2O[m] + pyruvate[m] => ADP[m] + oxaloacetate[m] + phosphate[m]\n",
      "\tpyruvate:carbon-dioxide ligase (ADP-forming)\n",
      "\tFlux: 0.16454 Reference flux: 0.70776 Difference: -0.54323\n",
      "\n",
      "r0115: diphosphate[m] + H2O[m] => 2 phosphate[m]\n",
      "\tdiphosphate phosphohydrolase\n",
      "\tFlux: 0 Reference flux: 1.8845 Difference: -1.8845\n",
      "\n",
      "r0118: 4 ferrocytochrome c[m] + 8 H(+) (energy metabolism)[m] + O2[m] => 8 H(+) (energy metabolism)[c] + 4 ferricytochrome c[m] + 2 H2O[m]\n",
      "\tferrocytochrome-c:oxygen oxidoreductase\n",
      "\tFlux: 2.9049 Reference flux: 5.0523 Difference: -2.1474\n",
      "\n",
      "r0119: 4 H(+) (energy metabolism)[m] + NADH[m] + ubiquinone[m] => 4 H(+) (energy metabolism)[c] + NAD(+)[m] + ubiquinol[m]\n",
      "\tNADH:ubiquinone oxidoreductase\n",
      "\tFlux: 3.0878 Reference flux: 5.8294 Difference: -2.7417\n",
      "\n",
      "r0127: 2 ferricytochrome c[m] + 4 H(+) (energy metabolism)[m] + ubiquinol[m] => 4 H(+) (energy metabolism)[c] + 2 ferrocytochrome c[m] + ubiquinone[m]\n",
      "\tubiquinol:ferricytochrome-c oxidoreductase\n",
      "\tFlux: 5.8098 Reference flux: 10.1046 Difference: -4.2947\n",
      "\n",
      "r0128: 3.88 H(+) (energy metabolism)[c] + ADP[m] + phosphate[m] => ATP[m] + 3.88 H(+) (energy metabolism)[m] + H2O[m]\n",
      "\tH+-transporting two-sector ATPase\n",
      "\tFlux: 13.4161 Reference flux: 23.5764 Difference: -10.1603\n",
      "\n",
      "r0377: 2 ADP[m] <=> AMP[m] + ATP[m]\n",
      "\tadenylate kinase\n",
      "\tFlux: 0 Reference flux: -1.8845 Difference: 1.8845\n",
      "\n",
      "r0542: 2-oxoglutarate[c] + L-alanine[c] <=> L-glutamate[c] + pyruvate[c]\n",
      "\tL-alanine:2-oxoglutarate aminotransferase\n",
      "\tFlux: -0.081019 Reference flux: -1.0932 Difference: 1.0122\n",
      "\n",
      "r0553: 2-oxoglutarate[c] + NADPH[c] + NH3[c] => H2O[c] + L-glutamate[c] + NADP(+)[c]\n",
      "\tL-glutamate:NADP+ oxidoreductase (deaminating)\n",
      "\tFlux: 0.4735 Reference flux: 1.5394 Difference: -1.0659\n",
      "\n",
      "r0558: glyoxylate[p] + L-alanine[p] <=> glycine[p] + pyruvate[p]\n",
      "\tL-alanine:glyoxylate aminotransferase\n",
      "\tFlux: 0.055199 Reference flux: 1.0602 Difference: -1.005\n",
      "\n",
      "r0625: L-threonine[c] <=> acetaldehyde[c] + glycine[c]\n",
      "\tL-threonine acetaldehyde-lyase (glycine-forming)\n",
      "\tFlux: -0.031173 Reference flux: -0.92089 Difference: 0.88972\n",
      "\n",
      "r0633: L-threonine[m] => 2-oxobutanoate[m] + NH3[m]\n",
      "\tL-threonine ammonia-lyase (2-oxobutanoate-forming)\n",
      "\tFlux: 0.013536 Reference flux: 0.8984 Difference: -0.88486\n",
      "\n",
      "r1170: 2-oxoglutarate[c] <=> 2-oxoglutarate[m]\n",
      "\t2-oxoglutarate mitochondrial permease\n",
      "\tFlux: 0.67604 Reference flux: 0 Difference: 0.67604\n",
      "\n",
      "r1213: L-threonine[m] <=> L-threonine[c]\n",
      "\tL-threonine mitochondrial permease\n",
      "\tFlux: -0.013536 Reference flux: -0.8984 Difference: 0.88486\n",
      "\n",
      "r1217: (S)-malate[c] + succinate[m] <=> succinate[c] + (S)-malate[m]\n",
      "\tmalate/succinate mitochondrial shuttle\n",
      "\tFlux: 0 Reference flux: -0.61335 Difference: 0.61335\n",
      "\n",
      "r1235: pyruvate[c] <=> pyruvate[m]\n",
      "\tpyruvate mitochondrial permease\n",
      "\tFlux: 1.5123 Reference flux: -0.030007 Difference: 1.5423\n",
      "\n",
      "r1243: acetaldehyde[c] <=> acetaldehyde[m]\n",
      "\tacetaldehyde mitochondrial membrane diffusion\n",
      "\tFlux: 0 Reference flux: 1.8845 Difference: -1.8845\n",
      "\n",
      "r1250: NH3[c] <=> NH3[m]\n",
      "\tNH3 mitochondrial membrane diffusion\n",
      "\tFlux: -0.013536 Reference flux: -0.95275 Difference: 0.93922\n",
      "\n",
      "r1251: O2[c] <=> O2[m]\n",
      "\tO2 mitochondrial membrane diffusion\n",
      "\tFlux: 2.905 Reference flux: 5.0523 Difference: -2.1474\n",
      "\n",
      "r1263: glycine[c] <=> glycine[p]\n",
      "\tglycine peroxisomal permease\n",
      "\tFlux: -0.055199 Reference flux: -1.0602 Difference: 1.005\n",
      "\n",
      "r1270: isocitrate[c] + (S)-malate[p] <=> (S)-malate[c] + isocitrate[p]\n",
      "\tisocitrate/malate peroxisomal permease\n",
      "\tFlux: 0.055182 Reference flux: 1.0602 Difference: -1.005\n",
      "\n",
      "r1273: L-alanine[c] <=> L-alanine[p]\n",
      "\tL-alanine peroxisomal permease\n",
      "\tFlux: 0.055199 Reference flux: 1.0602 Difference: -1.005\n",
      "\n",
      "r1276: (S)-malate[c] <=> (S)-malate[p]\n",
      "\tmalate peroxisomal permease\n",
      "\tFlux: 0.055182 Reference flux: 1.0602 Difference: -1.005\n",
      "\n",
      "r1284: pyruvate[c] <=> pyruvate[p]\n",
      "\tpyruvate peroxisomal permease\n",
      "\tFlux: -0.055182 Reference flux: -1.0602 Difference: 1.005\n",
      "\n",
      "r1286: succinate[c] <=> succinate[p]\n",
      "\tsuccinate peroxisomal permease\n",
      "\tFlux: -0.055182 Reference flux: -1.0602 Difference: 1.005\n",
      "\n",
      "r1316: alpha-D-glucose[e] <=> alpha-D-glucose[c]\n",
      "\talpha-D-glucose permease\n",
      "\tFlux: 1 Reference flux: 0 Difference: 1\n",
      "\n",
      "r1426: ethanol[c] <=> ethanol[e]\n",
      "\tethanol plasma membrane diffusion\n",
      "\tFlux: 0 Reference flux: -3 Difference: 3\n",
      "\n",
      "r1428: H2O[c] <=> H2O[e]\n",
      "\tH2O plasma membrane diffusion\n",
      "\tFlux: 5.586 Reference flux: 8.4719 Difference: -2.8859\n",
      "\n",
      "r1441: O2[e] <=> O2[c]\n",
      "\tO2 plasma membrane diffusion\n",
      "\tFlux: 2.937 Reference flux: 5.0932 Difference: -2.1562\n",
      "\n",
      "h2oOUT: H2O[e] => \n",
      "\tproduction of H2O\n",
      "\tFlux: 5.586 Reference flux: 8.4719 Difference: -2.8859\n",
      "\n",
      "glcIN:  => alpha-D-glucose[e]\n",
      "\tuptake of alpha-D-glucose\n",
      "\tFlux: 1 Reference flux: 0 Difference: 1\n",
      "\n",
      "o2IN:  => O2[e]\n",
      "\tuptake of O2\n",
      "\tFlux: 2.937 Reference flux: 5.0932 Difference: -2.1562\n",
      "\n",
      "etohIN:  => ethanol[e]\n",
      "\tuptake of ethanol\n",
      "\tFlux: 0 Reference flux: 3 Difference: -3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "% fluxesA: flux vector for the test case.fluxesB: flux vector for the reference test\n",
    "followChanged(modelETH,sol.x,solETH.x, 50, 0.5, 0.5);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These reactions have flux values that differ by more than 30 percent, absolute values above 0.4, and a total difference above 0.4 (10 reactions)\n",
      "\n",
      "Only prints reactions involving one or more of the following metabolites:\n",
      "ATP \n",
      "\n",
      "r0004: alpha-D-glucose[c] + ATP[c] => ADP[c] + alpha-D-glucose 6-phosphate[c]\n",
      "\tATP:alpha-D-glucose 6-phosphotransferase\n",
      "\tFlux: 1 Reference flux: 0 Difference: 1\n",
      "\n",
      "r0009: ATP[c] + beta-D-fructofuranose 6-phosphate[c] => ADP[c] + beta-D-fructofuranose 1,6-bisphosphate[c]\n",
      "\tphosphofructokinase\n",
      "\tFlux: 0.87408 Reference flux: 0 Difference: 0.87408\n",
      "\n",
      "r0013: 3-phospho-D-glyceroyl phosphate[c] + ADP[c] <=> 3-phospho-D-glycerate[c] + ATP[c]\n",
      "\tphosphoglycerate kinase\n",
      "\tFlux: 1.6507 Reference flux: -0.44547 Difference: 2.0962\n",
      "\n",
      "r0016: ADP[c] + phosphoenolpyruvate[c] => ATP[c] + pyruvate[c]\n",
      "\tpyruvate kinase\n",
      "\tFlux: 1.567 Reference flux: 0 Difference: 1.567\n",
      "\n",
      "r0018: ATP[c] + oxaloacetate[c] => ADP[c] + CO2[c] + phosphoenolpyruvate[c]\n",
      "\tphosphoenolpyruvate carboxykinase (ATP)\n",
      "\tFlux: 0 Reference flux: 0.4979 Difference: -0.4979\n",
      "\n",
      "r0026: acetate[m] + ATP[m] + coenzyme A[m] => acetyl-CoA[m] + AMP[m] + diphosphate[m]\n",
      "\tacetate:CoA ligase (AMP-forming)\n",
      "\tFlux: 0 Reference flux: 1.8845 Difference: -1.8845\n",
      "\n",
      "r0111: ATP[m] + CO2[m] + H2O[m] + pyruvate[m] => ADP[m] + oxaloacetate[m] + phosphate[m]\n",
      "\tpyruvate:carbon-dioxide ligase (ADP-forming)\n",
      "\tFlux: 0.16454 Reference flux: 0.70776 Difference: -0.54323\n",
      "\n",
      "r0128: 3.88 H(+) (energy metabolism)[c] + ADP[m] + phosphate[m] => ATP[m] + 3.88 H(+) (energy metabolism)[m] + H2O[m]\n",
      "\tH+-transporting two-sector ATPase\n",
      "\tFlux: 13.4161 Reference flux: 23.5764 Difference: -10.1603\n",
      "\n",
      "r0377: 2 ADP[m] <=> AMP[m] + ATP[m]\n",
      "\tadenylate kinase\n",
      "\tFlux: 0 Reference flux: -1.8845 Difference: 1.8845\n",
      "\n",
      "r1178: ADP[c] + H(+) (energy metabolism)[c] + phosphate[c] + ATP[m] + H2O[m] => ATP[c] + H2O[c] + ADP[m] + H(+) (energy metabolism)[m] + phosphate[m]\n",
      "\tADP/ATP/phosphate mitochondrial shuttle\n",
      "\tFlux: 13.9437 Reference flux: 19.4962 Difference: -5.5525\n",
      "\n"
     ]
    }
   ],
   "source": [
    "% You can filter using a metabolite list:\n",
    "followChanged(modelETH,sol.x,solETH.x, 30, 0.4, 0.4,{'ATP'});"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To understand better the underlying flux distributions the fluxes can be visualized in a map:\n",
    "\n",
    "- [load](https://www.mathworks.com/help/matlab/ref/load.html)<br>\n",
    "\n",
    "```MATLAB\n",
    "load(filename,variables)\n",
    "```\n",
    "Loads the specified variables from the MAT-file filename.\n",
    "\n",
    "- [drawMap](https://sysbiochalmers.github.io/RAVEN/doc/plotting/drawMap.html)<br>\n",
    "\n",
    "```MATLAB\n",
    "function notMapped=drawMap(title,pathway,modelA,conditionA,conditionB,modelB,filename,cutOff,supressOpen)\n",
    "```\n",
    "Imports a previously drawn map of the metabolic network and plots the fluxes on that map. If the pathway contains expression data the log-fold changes are plotted as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File saved as ./output/tutorial1_GLCvsETH.pdf\n"
     ]
    }
   ],
   "source": [
    "load('./tutorial_data/pcPathway.mat', \"pathway\");\n",
    "drawMap('Glucose vs ethanol',pathway,model,sol.x,solETH.x,modelETH,'./output/tutorial1_GLCvsETH.pdf',10^-5);"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "MATLAB Kernel",
   "language": "matlab",
   "name": "jupyter_matlab_kernel"
  },
  "language_info": {
   "file_extension": ".m",
   "mimetype": "text/x-matlab",
   "name": "matlab"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
